### PR TITLE
Refine DataTree.equals and DataTree.identical

### LINKED ISSUE #9215: DataTree.identical should consider checking that coordinates are defined at the same level
### What is your issue?

Consider the following two data trees:
```python
one = DataTree.from_dict(
    {
        "/": xr.Dataset(coords={"a": 1}),
        "/b": xr.Dataset(coords={"a": 1}),
    }
)
two = DataTree.from_dict(
    {
        "/": xr.Dataset(coords={"a": 1}),
        "/b": xr.Dataset(),
    }
)
```

Currently, `DataTree.identical()` considers `one` and `two` to be identical, but the coordinate `a` is defined redundantly in the first tree.

For most purposes, these DataTree objects are identical, but are edge cases where it could be different, e.g., when wrting these trees to netCDF or Zarr.

It would be helpful if the `.identical()` checked for such discrepancies, e.g., to facilitate writing unit tests as in https://github.com/pydata/xarray/pull/9214.

### PR DESCRIPTION
**Pull request by shoyer**
`.identical()` now checks for inherited coordinates being defined at the same level ("identical on disk"), whereas `.equals` now checks only that inherited coordinates are equal, even if defined redundantly ("good enough for xarray's alignment").

I've also removed the `from_root` argument from `identical` and `equals`, though we could bring it back if @TomNicholas feels strongly. It felt like a weird thing to check -- if you want to check from the root, why not just directly compare the root objects instead?

We should probably also check that `.path` matches, in the case of `.identical` only. Currently this is also checked for `.equals`, which I would propose we remove.

- [x] Closes #9215
- [x] Tests added

### CODE REVIEW

File: xarray/core/datatree.py [hunk 1]
```diff
@@ -1238,65 +1238,67 @@ def isomorphic(
         except (TypeError, TreeIsomorphismError):
             return False
 
-    def equals(self, other: DataTree, from_root: bool = True) -> bool:
+    def _matching(
+        self,
+        other: DataTree,
+        nodes_match: Callable[[DataTree, DataTree], bool],
+    ) -> bool:
+        if not self.isomorphic(other, from_root=True, strict_names=True):
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1246]: I'm wondering if there is a problem with removing the `from_root` argument but keeping it fixed to `True` inside `check_isomorphic`...

Let's imagines I `assert_equal` two subtrees that have different parents. All the nodes in the subtrees could match, but if the structures above the node where the subtree starts are not isomorphic, it will return `False`. That doesn't seem like desired behaviour to me - it leaves no way to check that the two subtrees are equal.

File: xarray/core/datatree.py [hunk 2]
```diff
@@ -1238,65 +1238,67 @@ def isomorphic(
...
 
         See Also
         --------
         Dataset.identical
         DataTree.isomorphic
         DataTree.equals
         """
-        if not self.isomorphic(other, from_root=from_root, strict_names=True):
+        # include inherited coordinates only at the root; otherwise require that
+        # everything is also defined at the same level
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1294]: Why do this for `identical` but not for `equals`?

File: xarray/core/datatree.py [hunk 3]
```diff
@@ -1238,65 +1238,67 @@ def isomorphic(
...
-            If neither tree has a parent then this has no effect.
 
         See Also
         --------
         Dataset.equals
         DataTree.isomorphic
         DataTree.identical
         """
-        if not self.isomorphic(other, from_root=from_root, strict_names=True):
+        to_ds = lambda x: x._to_dataset_view(inherited=True, rebuild_dims=False)
```
Comment (TomNicholas) [inline xarray/core/datatree.py:1270]: Wouldn't it be simpler (and clearer) to use `inherited=False`? The nodes are checked in depth-first order anyway, so if there is a discrepancy between inherited coordinates it will be discovered at the top of the subtree before checking the children.

**Comment by TomNicholas** [comment]
> It felt like a weird thing to check -- if you want to check from the root, why not just directly compare the root objects instead?

I feel like maybe there was some reason why I needed this, in `map_over_subtree` maybe?

File: xarray/core/datatree.py (hunk 2 above)
Comment (TomNicholas) [reply xarray/core/datatree.py:1294]: Oh you did talk about this in your comment above

**Comment by TomNicholas** [comment]
> I feel like maybe there was some reason why I needed this, in `map_over_subtree` maybe?

`map_over_subtree` only uses `from_root=False`, so I guess we can get rid of this argument 🤷 https://github.com/pydata/xarray/blob/cea354fbfb5f3ddeeaf7de5fbf9558290eaf154b/xarray/core/datatree_mapping.py#L141

**Comment by TomNicholas** [comment]
Closing this based on the comment in https://github.com/pydata/xarray/issues/9215#issuecomment-2403674406. Feel free to re-open if you think we do want to keep some part of this PR.
